core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridImpl.java

```diff
@@ -0,0 +1,118 @@
+                                // Check if the queue size is more than the actual
+                                // no of browsers allowed per worker
+                                // means crawler services are idle because all drivers are in queue
+                                if (driversQueue.size() >= (size / noOfWorkers)) {
+                                    SeleniumGridProtocol.Holder holder = driversQueue.take();
+                                    long totalTime = System.currentTimeMillis() - holder.getTime();
+                                    // clear the queue if the total time spent in the queue is more
+                                    // than 4.5 minutes
+                                    // As after 5 mintues the driver would be idle and will throw
+                                    // exception
+                                    // while we try to use that driver for fetching
+                                    if (totalTime > 1000 * 4.5 * 60) {
+                                        driversQueue.clear();
+                                    }
+                                }
+                                // so that browsers get equally divided among workers
+                                if (driversQueue.size() <= size / noOfWorkers) {
+                                    RemoteWebDriver driver = getDriverFromNode();
+                                    if (driver != null) {
+                                        driversQueue.put(
+                                                new SeleniumGridProtocol.Holder(
+                                                        driver, System.currentTimeMillis()));
+                                        LOG.info(
+                                                "Placed driver in blocking queue: "
+                                                        + driversQueue.size());
+                                    }
+                                }
+                                list = getAllNodesList();
+                            }
+                        } catch (Exception e) {
+                            LOG.error(
+                                    "Exception while running task for adding driver to the queue",
+                                    e);
+                        }
+                    }
+                };
+        // update the queue every 5 minutes
+        updateQueue.schedule(timerTask, 0 * 60 * 1000, 5 * 60 * 1000);
+    }
+
+    protected RemoteWebDriver getDriverFromNode() {
+        int sessionCount = 0;
+        RemoteWebDriver driver = null;
+
+        try {
+            LOG.debug("Adding new driver from " + gridAddress);
+            driver = new RemoteWebDriver(new URL(gridAddress), capabilities);
+            WebDriver.Timeouts touts = driver.manage().timeouts();
+            WebDriver.Window window = driver.manage().window();
+            touts.implicitlyWait(Duration.ofSeconds(6));
```

Comment:
can we add configuration options for these hard-coded values? I think, that we already have them?

Reply:
sure will do it